gunicorn/reloader.py

```diff
@@ -66,49 +68,45 @@ def run(self):
 
 if has_inotify:
 
-    class InotifyReloader(threading.Thread):
+    class InotifyReloader(ReloaderBase):
         event_mask = (inotify.constants.IN_CREATE | inotify.constants.IN_DELETE
                       | inotify.constants.IN_DELETE_SELF | inotify.constants.IN_MODIFY
                       | inotify.constants.IN_MOVE_SELF | inotify.constants.IN_MOVED_FROM
                       | inotify.constants.IN_MOVED_TO)
 
         def __init__(self, extra_files=None, callback=None):
-            super().__init__()
-            self.setDaemon(True)
-            self._callback = callback
+            super().__init__(extra_files=extra_files, callback=callback)
             self._dirs = set()
             self._watcher = Inotify()
 
-            for extra_file in extra_files:
-                self.add_extra_file(extra_file)
-
         def add_extra_file(self, filename):
             dirname = os.path.dirname(filename)
 
+            super(InotifyReloader, self).add_extra_file(filename)
+
             if dirname in self._dirs:
                 return
 
             self._watcher.add_watch(dirname, mask=self.event_mask)
             self._dirs.add(dirname)
 
         def get_dirs(self):
-            fnames = [
-                os.path.dirname(os.path.abspath(COMPILED_EXT_RE.sub('py', module.__file__)))
-                for module in tuple(sys.modules.values())
-                if getattr(module, '__file__', None)
-            ]
+            dirnames = [os.path.dirname(os.path.abspath(fname)) for fname in self.get_files()]
+            return set(dirnames)
 
-            return set(fnames)
+        def refresh_dirs(self):
+            new_dirs = self.get_dirs().difference(self._dirs)
+            self._dirs.update(new_dirs)
+            for new_dir in new_dirs:
+                if os.path.isdir(new_dir):
+                    self._watcher.add_watch(new_dir, mask=self.event_mask)
 
         def run(self):
-            self._dirs = self.get_dirs()
-
-            for dirname in self._dirs:
-                if os.path.isdir(dirname):
-                    self._watcher.add_watch(dirname, mask=self.event_mask)
+            self.refresh_dirs()
 
             for event in self._watcher.event_gen():
                 if event is None:
+                    self.refresh_dirs()
```

Comment:
Why is this the right place to call this?

Reply:
Maybe this is the key I'm missing. I don't think I understand how this fixes the issue with Django because I don't know when the watcher would generate an empty event and cause this refresh. When will the reloader know to refresh and what happens if the file that changes is not in the initial set of directories that Django imported from?

Reply:
By default, `Inotify()` will watch for one second at a time, so this line will end up refreshing the list of watched directories every second.